File: bindings/csharp/Builders.cs
// Handwritten endpoint builders for the transport.rest C# binding.
using System.Globalization;
using System.Text.Json;
using TransportRest.Models;

namespace TransportRest;

public static class BuilderUtil
{
    public static string B(bool v) => v ? "true" : "false";
    public static string Num(double v) => v.ToString(CultureInfo.InvariantCulture);
    public static string Iso(DateTimeOffset when) => when.ToString("yyyy-MM-dd'T'HH:mm:sszzz", CultureInfo.InvariantCulture);
}

public sealed class ProductSelection
{
    private readonly List<(string Key, bool Enabled)> entries = new();

    public ProductSelection Set(string key, bool enabled)
    {
        for (var i = 0; i < entries.Count; i++)
            if (entries[i].Key == key) { entries[i] = (key, enabled); return this; }
        entries.Add((key, enabled));
        return this;
    }

    public ProductSelection NationalExpress(bool v) => Set("nationalExpress", v);
    public ProductSelection National(bool v) => Set("national", v);
    public ProductSelection RegionalExpress(bool v) => Set("regionalExpress", v);
    public ProductSelection Regional(bool v) => Set("regional", v);
    public ProductSelection Suburban(bool v) => Set("suburban", v);
    public ProductSelection Subway(bool v) => Set("subway", v);
    public ProductSelection Tram(bool v) => Set("tram", v);
    public ProductSelection Bus(bool v) => Set("bus", v);
    public ProductSelection Ferry(bool v) => Set("ferry", v);
    public ProductSelection Taxi(bool v) => Set("taxi", v);
    public ProductSelection Express(bool v) => Set("express", v);

    internal void Encode(List<(string, string)> parameters)
    {
        foreach (var (key, enabled) in entries)
            parameters.Add((key, enabled ? "true" : "false"));
    }
}

internal abstract class BuilderBase<T>
{
    protected readonly TransportRestClient client;
    private readonly Func<JsonElement, T> parse;
    private readonly string path;
    private readonly string? capability;

    protected BuilderBase(TransportRestClient client, string path,
        Func<JsonElement, T> parse, string? capability = null)
    {
        this.client = client;
        this.path = path;
        this.parse = parse;
        this.capability = capability;
    }

    private List<(string, string)> current = new();
    protected List<(string, string)> Parameters { get; } = new();

    protected void SetParameters(List<(string, string)> replacement)
    {
        Parameters.Clear();
        Parameters.AddRange(replacement);
    }

    protected void Require(bool condition, string? parameter, string reason)
    {
        if (!condition) throw new InvalidParameterException(parameter, reason);
    }

    public Task<T> GetAsync(CancellationToken cancellationToken = default) =>
        client.GetJsonAsync(path, Parameters, parse, capability, cancellationToken);

}

public sealed class LocationsBuilder : BuilderBase<IReadOnlyList<LocationResult>>
{
    private string? query;

    internal LocationsBuilder(TransportRestClient client)
        : base(client, "/locations", d => JsonSerializer.Deserialize<IReadOnlyList<LocationResult>>(d)!) { }

    public LocationsBuilder Query(string q)
    {
        Require(!string.IsNullOrWhiteSpace(q), "query", "a non-empty search term is required");
        query = q;
        return this;
    }

    public LocationsBuilder Results(int n) { Parameters.Add(("results", n.ToString())); return this; }
    public LocationsBuilder Fuzzy(bool v) { Parameters.Add(("fuzzy", BuilderUtil.B(v))); return this; }
    public LocationsBuilder Language(string l) { Parameters.Add(("language", l)); return this; }

    public new Task<IReadOnlyList<LocationResult>> GetAsync(CancellationToken ct = default)
    {
        Require(query is not null, "query", "Query() is required before GetAsync()");
        Parameters.Insert(0, ("query", query!));
        return base.GetAsync(ct);
    }
}

public abstract class BoardBuilder<T> : BuilderBase<T>
{
    protected BoardBuilder(TransportRestClient client, string path, Func<JsonElement, T> parse)
        : base(client, path, parse) { }

    public BoardBuilder<T> When(DateTimeOffset when) { Parameters.Add(("when", BuilderUtil.Iso(when))); return this; }
    public BoardBuilder<T> Direction(string d) { Parameters.Add(("direction", d)); return this; }
    public BoardBuilder<T> Duration(int minutes) { Parameters.Add(("duration", minutes.ToString())); return this; }
    public BoardBuilder<T> Results(int n) { Parameters.Add(("results", n.ToString())); return this; }
    public BoardBuilder<T> Stopovers(bool v) { Parameters.Add(("stopovers", BuilderUtil.B(v))); return this; }
    public BoardBuilder<T> Remarks(bool v) { Parameters.Add(("remarks", BuilderUtil.B(v))); return this; }
    public BoardBuilder<T> MoreStops(IEnumerable<string> ids) { Parameters.Add(("moreStops", string.Join(",", ids))); return this; }

    public BoardBuilder<T> Products(Func<ProductSelection, ProductSelection> configure)
    {
        configure(new ProductSelection()).Encode(Parameters);
        return this;
    }
}

public sealed class DeparturesBuilder : BoardBuilder<DeparturesResponse>
{
    internal static DeparturesBuilder Departures(TransportRestClient c, string stopId) =>
        new(c, $"/stops/{Uri.EscapeDataString(stopId)}/departures");

    internal static ArrivalsAdapter Arrivals(TransportRestClient c, string stopId) =>
        new(c, $"/stops/{Uri.EscapeDataString(stopId)}/arrivals");

    private DeparturesBuilder(TransportRestClient client, string path)
        : base(client, path,
            d => JsonSerializer.Deserialize<DeparturesResponse>(d)!)
    {
        Require(!string.IsNullOrWhiteSpace(stopId), "stop_id", "must not be empty");
    }
}

public sealed class ArrivalsAdapter
{
    private readonly TransportRestClient client;
    private readonly string path;

    internal ArrivalsAdapter(TransportRestClient client, string path)
    {
        this.client = client;
        this.path = path;
    }

    public Task<ArrivalsResponse> GetAsync(CancellationToken ct = default) =>
        client.GetJsonAsync(path, new List<(string, string)>(),
            d => JsonSerializer.Deserialize<ArrivalsResponse>(d)!);
}

public sealed class JourneysBuilder : BuilderBase<JourneysResponse>
{
    private readonly JourneyPlace from;
    private readonly JourneyPlace to;
    private JourneyPlace? via;

    internal JourneysBuilder(TransportRestClient client, JourneyPlace from, JourneyPlace to)
        : base(client, "/journeys", d => JsonSerializer.Deserialize<JourneysResponse>(d)!)
    {
        this.from = from;
        this.to = to;
    }

    public JourneysBuilder Via(JourneyPlace place) { via = place; return this; }

    public JourneysBuilder Departure(DateTimeOffset when) { Parameters.Add(("departure", BuilderUtil.Iso(when))); return this; }
    public JourneysBuilder Arrival(DateTimeOffset when) { Parameters.Add(("arrival", BuilderUtil.Iso(when))); return this; }
    public JourneysBuilder EarlierThan(string reference) { Parameters.Add(("earlierThan", reference)); return this; }
    public JourneysBuilder LaterThan(string reference) { Parameters.Add(("laterThan", reference)); return this; }
    public JourneysBuilder Results(int n) { Parameters.Add(("results", n.ToString())); return this; }
    public JourneysBuilder Transfers(int n) { Parameters.Add(("transfers", n.ToString())); return this; }

    public JourneysBuilder Products(Func<ProductSelection, ProductSelection> configure)
    {
        configure(new ProductSelection()).Encode(Parameters);
        return this;
    }

    public new Task<JourneysResponse> GetAsync(CancellationToken ct = default)
    {
        from.Validate("from");
        to.Validate("to");
        via?.Validate("via");
        var params2 = new List<(string, string)>();
        from.Encode("from", params2);
        to.Encode("to", params2);
        via?.Encode("via", params2);
        params2.AddRange(Parameters);
        SetParameters(params2);
        return base.GetAsync(ct);
    }
}

public sealed class TripBuilder : BuilderBase<TripResponse>
{
    internal TripBuilder(TransportRestClient client, string tripId)
        : base(client, $"/trips/{Uri.EscapeDataString(tripId)}",
            d => JsonSerializer.Deserialize<TripResponse>(d)!)
    {
        Require(!string.IsNullOrWhiteSpace(tripId), "trip_id", "must not be empty");
    }

    public TripBuilder Stopovers(bool v) { Parameters.Add(("stopovers", BuilderUtil.B(v))); return this; }
    public TripBuilder Remarks(bool v) { Parameters.Add(("remarks", BuilderUtil.B(v))); return this; }
    public TripBuilder Polyline(bool v) { Parameters.Add(("polyline", BuilderUtil.B(v))); return this; }
}

public sealed class RadarBuilder : BuilderBase<RadarResponse>
{
    private readonly Dictionary<string, double> box = new();

    internal RadarBuilder(TransportRestClient client)
        : base(client, "/radar", d => JsonSerializer.Deserialize<RadarResponse>(d)!, "radar") { }

    public RadarBuilder North(double v) { box["north"] = v; return this; }
    public RadarBuilder West(double v) { box["west"] = v; return this; }
    public RadarBuilder South(double v) { box["south"] = v; return this; }
    public RadarBuilder East(double v) { box["east"] = v; return this; }
    public RadarBuilder Results(int n) { Parameters.Add(("results", n.ToString())); return this; }

    public new Task<RadarResponse> GetAsync(CancellationToken ct = default)
    {
        Require(box.Count == 4, "bbox", "north, west, south and east are all required");
        foreach (var (key, value) in box.OrderBy(kv => kv.Key))
            Parameters.Add((key, BuilderUtil.Num(value)));
        return base.GetAsync(ct);
    }
}

File: bindings/typescript/src/builders.ts
/** Endpoint builders mirroring the Rust core API. */

import type { TransportRestClient, Capability, ModelParser } from "./client.js";
import { ApiError } from "./errors.js";
import {
  bool as b,
  iso,
  require,
  JourneyPlace,
  encodePathSegment,
  type QueryParam,
} from "./util.js";
import type {
  ArrivalsResponse,
  DeparturesResponse,
  JourneysResponse,
  JourneyResponse,
  LocationResult,
  RadarResponse,
  ReachableFromResponse,
  Station,
  StopOrStation,
  TripResponse,
  TripsResponse,
} from "./models.js";

export { JourneyPlace };

function parseLocations(data: unknown): LocationResult[] {
  if (!Array.isArray(data)) throw new TypeError("expected array of locations");
  return data as LocationResult[];
}

function parseStations(data: unknown): Station[] {
  if (!Array.isArray(data)) throw new TypeError("expected array of stations");
  return data as Station[];
}

export class Builder<T> {
  constructor(
    protected readonly client: TransportRestClient,
    protected readonly path: string,
    private readonly model: ModelParser<T>,
    protected readonly capability?: Capability,
  ) {}

  protected params: QueryParam[] = [];

  protected opt(key: string, value: string | number | undefined): void {
    if (value !== undefined) this.params.push([key, String(value)]);
  }

  get(): Promise<T> {
    return this.client.get_json(this.path, this.params, this.model, this.capability);
  }
}

export class LocationsBuilder extends Builder<LocationResult[]> {
  constructor(client: TransportRestClient) {
    super(client, "/locations", parseLocations);
  }

  query(q: string): this {
    require(q && q.trim(), "query", "a non-empty search term is required");
    this.params.unshift(["query", q]);
    return this;
  }
  fuzzy(value: boolean): this {
    this.opt("fuzzy", b(value));
    return this;
  }
  results(n: number): this {
    this.opt("results", n);
    return this;
  }
  stops(value: boolean): this {
    this.opt("stops", b(value));
    return this;
  }
  addresses(value: boolean): this {
    this.opt("addresses", b(value));
    return this;
  }
  poi(value: boolean): this {
    this.opt("poi", b(value));
    return this;
  }
  linesOfStops(value: boolean): this {
    this.opt("linesOfStops", b(value));
    return this;
  }
  language(language: string): this {
    this.opt("language", language);
    return this;
  }

  override async get(): Promise<LocationResult[]> {
    const q = this.params.find(([k]) => k === "query");
    require(q !== undefined, undefined, "query() is required before get()");
    return await super.get();
  }
}

export class NearbyBuilder extends Builder<LocationResult[]> {
  private lat?: number;
  private lon?: number;

  constructor(client: TransportRestClient) {
    super(client, "/locations/nearby", parseLocations);
  }

  latitude(v: number): this {
    require(v >= -90 && v <= 90, "latitude", "must be within [-90, 90]");
    this.lat = v;
    return this;
  }
  longitude(v: number): this {
    require(v >= -180 && v <= 180, "longitude", "must be within [-180, 180]");
    this.lon = v;
    return this;
  }
  results(n: number): this {
    this.opt("results", n);
    return this;
  }
  distance(meters: number): this {
    this.opt("distance", meters);
    return this;
  }
  stops(value: boolean): this {
    this.opt("stops", b(value));
    return this;
  }
  poi(value: boolean): this {
    this.opt("poi", b(value));
    return this;
  }
  language(language: string): this {
    this.opt("language", language);
    return this;
  }

  override async get(): Promise<LocationResult[]> {
    require(
      this.lat !== undefined && this.lon !== undefined,
      undefined,
      "latitude and longitude are both required",
    );
    this.params.unshift(["longitude", String(this.lon)]);
    this.params.unshift(["latitude", String(this.lat)]);
    return await super.get();
  }
}

export class StopBuilder extends Builder<StopOrStation> {
  constructor(client: TransportRestClient, id: string) {
    super(client, `/stops/${encodePathSegment(id)}`, (d) => d as StopOrStation);
    require(id && id.trim(), "id", "stop ID must not be empty");
  }

  linesOfStops(value: boolean): this {
    this.opt("linesOfStops", b(value));
    return this;
  }
  language(language: string): this {
    this.opt("language", language);
    return this;
  }
}

interface BoardState {
  when?: Date;
  direction?: string;
  duration?: number;
  results?: number;
  stopovers?: boolean;
  includeRelatedStations?: boolean;
  linesOfStops?: boolean;
  remarks?: boolean;
  language?: string;
  moreStops?: string[];
  products?: ProductSelection;
}

export class ProductSelection {
  private readonly entries: QueryParam[] = [];

  set(key: string, enabled: boolean): this {
    const existing = this.entries.findIndex(([k]) => k === key);
    if (existing >= 0) this.entries[existing] = [key, b(enabled)];
    else this.entries.push([key, b(enabled)]);
    return this;
  }
  nationalExpress(v: boolean): this { return this.set("nationalExpress", v); }
  national(v: boolean): this { return this.set("national", v); }
  regionalExpress(v: boolean): this { return this.set("regionalExpress", v); }
  regional(v: boolean): this { return this.set("regional", v); }
  suburban(v: boolean): this { return this.set("suburban", v); }
  subway(v: boolean): this { return this.set("subway", v); }
  tram(v: boolean): this { return this.set("tram", v); }
  bus(v: boolean): this { return this.set("bus", v); }
  ferry(v: boolean): this { return this.set("ferry", v); }
  taxi(v: boolean): this { return this.set("taxi", v); }
  express(v: boolean): this { return this.set("express", v); }

  encodeInto(params: QueryParam[]): void {
    params.push(...this.entries);
  }
}

export class DeparturesBuilder extends Builder<DeparturesResponse | ArrivalsResponse> {
  private readonly board: BoardState = {};
  private productsSelection?: ProductSelection;

  static create(
    client: TransportRestClient,
    stopId: string,
    kind: "departures" | "arrivals",
  ): DeparturesBuilder {
    require(stopId && stopId.trim(), "stopId", "must not be empty");
    const model: ModelParser<DeparturesResponse> =
      kind === "departures"
        ? (d) => d as DeparturesResponse
        : (d) => d as unknown as DeparturesResponse;
    return new DeparturesBuilder(
      client,
      `/stops/${encodePathSegment(stopId)}/${kind}`,
      model,
    );
  }

  private constructor(client: TransportRestClient, path: string, model: ModelParser<DeparturesResponse>) {
    super(client, path, model);
  }

  when(when: Date): this {
    this.board.when = when;
    return this;
  }
  direction(direction: string): this {
    this.board.direction = direction;
    return this;
  }
  duration(minutes: number): this {
    this.board.duration = minutes;
    return this;
  }
  results(n: number): this {
    this.board.results = n;
    return this;
  }
  stopovers(value: boolean): this {
    this.board.stopovers = value;
    return this;
  }
  includeRelatedStations(value: boolean): this {
    this.board.includeRelatedStations = value;
    return this;
  }
  linesOfStops(value: boolean): this {
    this.board.linesOfStops = value;
    return this;
  }
  remarks(value: boolean): this {
    this.board.remarks = value;
    return this;
  }
  language(language: string): this {
    this.board.language = language;
    return this;
  }
  moreStops(stopIds: string[]): this {
    this.board.moreStops = stopIds;
    return this;
  }
  products(configure: (p: ProductSelection) => ProductSelection): this {
    this.productsSelection = configure(new ProductSelection());
    return this;
  }

  override async get(): Promise<DeparturesResponse> {
    if (this.board.when) this.params.unshift(["when", iso(this.board.when)]);
    if (this.board.direction !== undefined) this.params.unshift(["direction", this.board.direction]);
    if (this.board.duration !== undefined) this.params.unshift(["duration", String(this.board.duration)]);
    if (this.board.results !== undefined) this.params.unshift(["results", String(this.board.results)]);
    if (this.board.stopovers !== undefined) this.params.unshift(["stopovers", b(this.board.stopovers)]);
    if (this.board.includeRelatedStations !== undefined)
      this.params.unshift(["includeRelatedStations", b(this.board.includeRelatedStations)]);
    if (this.board.linesOfStops !== undefined)
      this.params.unshift(["linesOfStops", b(this.board.linesOfStops)]);
    if (this.board.remarks !== undefined) this.params.unshift(["remarks", b(this.board.remarks)]);
    if (this.board.language !== undefined) this.params.unshift(["language", this.board.language]);
    if (this.board.moreStops?.length) this.params.unshift(["moreStops", this.board.moreStops.join(",")]);
    this.productsSelection?.encodeInto(this.params);
    return (await super.get()) as DeparturesResponse;
  }
}

export class JourneysBuilder extends Builder<JourneysResponse> {
  // NOTE: field names deliberately avoid clashing with the fluent method
  // names below (useDefineForClassFields would otherwise shadow them).
  private readonly _from: JourneyPlace;
  private readonly _to: JourneyPlace;
  private _via?: JourneyPlace;
  private _departure?: Date;
  private _arrival?: Date;
  private _earlierThan?: string;
  private _laterThan?: string;
  private readonly options: QueryParam[] = [];
  private productsSelection?: ProductSelection;

  constructor(client: TransportRestClient, from: JourneyPlace, to: JourneyPlace) {
    super(client, "/journeys", (d) => d as JourneysResponse);
    this._from = from;
    this._to = to;
  }

  via(place: JourneyPlace): this {
    this._via = place;
    return this;
  }
  departure(when: Date): this {
    this._departure = when;
    return this;
  }
  arrival(when: Date): this {
    this._arrival = when;
    return this;
  }
  earlierThan(ref: string): this {
    this._earlierThan = ref;
    return this;
  }
  laterThan(ref: string): this {
    this._laterThan = ref;
    return this;
  }
  results(n: number): this { this.options.push(["results", String(n)]); return this; }
  stopovers(v: boolean): this { this.options.push(["stopovers", b(v)]); return this; }
  transfers(n: number): this { this.options.push(["transfers", String(n)]); return this; }
  transferTime(minutes: number): this { this.options.push(["transferTime", String(minutes)]); return this; }
  accessibility(value: "partial" | "complete"): this {
    this.options.push(["accessibility", value]);
    return this;
  }
  bike(v: boolean): this { this.options.push(["bike", b(v)]); return this; }
  startWithWalking(v: boolean): this { this.options.push(["startWithWalking", b(v)]); return this; }
  walkingSpeed(speed: "slow" | "normal" | "fast"): this {
    this.options.push(["walkingSpeed", speed]);
    return this;
  }
  tickets(v: boolean): this { this.options.push(["tickets", b(v)]); return this; }
  polylines(v: boolean): this { this.options.push(["polylines", b(v)]); return this; }
  remarks(v: boolean): this { this.options.push(["remarks", b(v)]); return this; }
  scheduledDays(v: boolean): this { this.options.push(["scheduledDays", b(v)]); return this; }
  notOnlyFastRoutes(v: boolean): this { this.options.push(["notOnlyFastRoutes", b(v)]); return this; }
  bestprice(v: boolean): this { this.options.push(["bestprice", b(v)]); return this; }
  loyaltyCard(card: string): this { this.options.push(["loyaltyCard", card]); return this; }
  firstClass(v: boolean): this { this.options.push(["firstClass", b(v)]); return this; }
  routingMode(mode: string): this { this.options.push(["routingMode", mode]); return this; }
  products(configure: (p: ProductSelection) => ProductSelection): this {
    this.productsSelection = configure(new ProductSelection());
    return this;
  }

  override async get(): Promise<JourneysResponse> {
    this._from.validate("from");
    this._to.validate("to");
    this._via?.validate("via");
    require(
      !(this._departure && this._arrival),
      undefined,
      "departure and arrival are mutually exclusive",
    );
    require(
      !((this._earlierThan || this._laterThan) && (this._departure || this._arrival)),
      undefined,
      "earlierThan/laterThan cannot be combined with departure/arrival",
    );
    require(
      !(this._earlierThan && this._laterThan),
      undefined,
      "earlierThan and laterThan are mutually exclusive",
    );

    const params: QueryParam[] = [];
    this._from.encode("from", params);
    this._to.encode("to", params);
    if (this._via) this._via.encode("via", params);
    if (this._departure) params.push(["departure", iso(this._departure)]);
    if (this._arrival) params.push(["arrival", iso(this._arrival)]);
    if (this._earlierThan) params.push(["earlierThan", this._earlierThan]);
    if (this._laterThan) params.push(["laterThan", this._laterThan]);
    params.push(...this.options);
    this.productsSelection?.encodeInto(params);
    this.params = params;
    return await super.get();
  }
}

export class RefreshJourneyBuilder extends Builder<JourneyResponse> {
  private ticketsValue?: boolean;
  private polylinesValue?: boolean;

  constructor(client: TransportRestClient, refreshToken: string) {
    super(
      client,
      `/journeys/${encodePathSegment(refreshToken)}`,
      (d) => d as JourneyResponse,
    );
    require(refreshToken && refreshToken.trim(), "refreshToken", "must not be empty");
  }

  stopovers(v: boolean): this { this.opt("stopovers", b(v)); return this; }
  tickets(v: boolean): this { this.ticketsValue = v; return this; }
  polylines(v: boolean): this { this.polylinesValue = v; return this; }
  remarks(v: boolean): this { this.opt("remarks", b(v)); return this; }
  language(language: string): this { this.opt("language", language); return this; }

  override async get(): Promise<JourneyResponse> {
    require(
      !(this.ticketsValue && this.polylinesValue),
      undefined,
      "tickets and polylines are mutually exclusive",
    );
    return await super.get();
  }
}

export class TripBuilder extends Builder<TripResponse> {
  constructor(client: TransportRestClient, id: string) {
    super(client, `/trips/${encodePathSegment(id)}`, (d) => d as TripResponse);
    require(id && id.trim(), "id", "trip ID must not be empty");
  }

  stopovers(v: boolean): this { this.opt("stopovers", b(v)); return this; }
  remarks(v: boolean): this { this.opt("remarks", b(v)); return this; }
  polyline(v: boolean): this { this.opt("polyline", b(v)); return this; }
  language(language: string): this { this.opt("language", language); return this; }
}

export class TripsByNameBuilder extends Builder<TripsResponse> {
  constructor(client: TransportRestClient, query: string) {
    super(client, "/trips", (d) => d as TripsResponse, "trips_by_name");
    this.opt("query", query || "*");
  }

  onlyCurrentlyRunning(v: boolean): this { this.opt("onlyCurrentlyRunning", b(v)); return this; }
  lineName(name: string): this { this.opt("lineName", name); return this; }
}

export class RadarBuilder extends Builder<RadarResponse> {
  private box: Partial<Record<"north" | "west" | "south" | "east", number>> = {};

  constructor(client: TransportRestClient) {
    super(client, "/radar", (d) => d as RadarResponse, "radar");
  }

  north(v: number): this { this.box.north = v; return this; }
  west(v: number): this { this.box.west = v; return this; }
  south(v: number): this { this.box.south = v; return this; }
  east(v: number): this { this.box.east = v; return this; }
  results(n: number): this { this.opt("results", n); return this; }
  frames(n: number): this { this.opt("frames", n); return this; }
  duration(seconds: number): this { this.opt("duration", seconds); return this; }

  override async get(): Promise<RadarResponse> {
    const box = this.box;
    require(
      box.north !== undefined && box.west !== undefined &&
        box.south !== undefined && box.east !== undefined,
      undefined,
      "north, west, south and east are all required",
    );
    require(
      (box.south as number) <= (box.north as number) &&
        (box.west as number) <= (box.east as number),
      undefined,
      "bounding box is invalid: require south <= north and west <= east",
    );
    const ordered: QueryParam[] = [
      ["north", String(box.north)],
      ["west", String(box.west)],
      ["south", String(box.south)],
      ["east", String(box.east)],
    ];
    this.params = [...ordered, ...this.params];
    return (await super.get()) as RadarResponse;
  }
}

export class ReachableFromBuilder extends Builder<ReachableFromResponse> {
  private lat?: number;
  private lon?: number;

  constructor(client: TransportRestClient) {
    super(client, "/stops/reachable-from", (d) => d as ReachableFromResponse, "reachable_from");
  }

  latitude(v: number): this { this.lat = v; return this; }
  longitude(v: number): this { this.lon = v; return this; }
  maxTransfers(n: number): this { this.opt("maxTransfers", n); return this; }
  maxDuration(minutes: number): this { this.opt("maxDuration", minutes); return this; }

  override async get(): Promise<ReachableFromResponse> {
    require(this.lat !== undefined && this.lon !== undefined, undefined, "latitude and longitude are both required");
    this.params.unshift(["longitude", String(this.lon)]);
    this.params.unshift(["latitude", String(this.lat)]);
    return (await super.get()) as ReachableFromResponse;
  }
}

export class StationsBuilder extends Builder<Station[]> {
  constructor(client: TransportRestClient) {
    super(client, "/stations", parseStations, "stations");
  }

  query(q: string): this { this.opt("query", q); return this; }
  results(n: number): this { this.opt("results", n); return this; }
}

export class StationBuilder extends Builder<Station> {
  constructor(client: TransportRestClient, id: string) {
    super(client, `/stations/${encodePathSegment(id)}`, (d) => d as Station, "stations");
    require(id && id.trim(), "id", "station ID must not be empty");
  }
}

export class StopsSearchBuilder extends Builder<LocationResult[]> {
  constructor(client: TransportRestClient) {
    super(client, "/stops", parseLocations, "stops_search");
  }

  query(q: string): this { this.opt("query", q); return this; }
  limit(n: number): this { this.opt("limit", n); return this; }
}
